feat: Subscription CRUD

navidshaikh opened the pull request:
## Description

Subscription CRUD operations

## Changes
- Add subscription create, update, list, describe delete commands
- Channel type mapping can be used while creating the subscription to refer different channel types
-  Set default channel type `messaging.knative.dev/v1beta1:Channel`, i.e. if no prefix is given to `--channel`, consider it of `Channel` type.


/lint

Fixes #955

mattmoor commented on pkg/kn/commands/subscription/create.go:
```diff
@@ -0,0 +1,110 @@
+// Copyright © 2020 The Knative Authors
```
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2020 The Knative Authors
```

mattmoor replied:
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2020 The Knative Authors
```

mattmoor replied:
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2020 The Knative Authors
```

mattmoor commented on pkg/kn/commands/subscription/create_test.go:
```diff
@@ -0,0 +1,61 @@
+// Copyright © 2020 The Knative Authors
```
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2020 The Knative Authors
```

mattmoor replied:
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2020 The Knative Authors
```

mattmoor commented on pkg/kn/commands/subscription/delete.go:
```diff
@@ -0,0 +1,55 @@
+// Copyright © 2020 The Knative Authors
```
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2020 The Knative Authors
```

mattmoor replied:
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2020 The Knative Authors
```

mattmoor replied:
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2020 The Knative Authors
```

mattmoor commented on pkg/kn/commands/subscription/delete_test.go:
```diff
@@ -0,0 +1,52 @@
+// Copyright © 2020 The Knative Authors
```
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2020 The Knative Authors
```

mattmoor replied:
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2020 The Knative Authors
```

mattmoor commented on pkg/kn/commands/subscription/describe.go:
```diff
@@ -0,0 +1,127 @@
+// Copyright © 2020 The Knative Authors
```
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2020 The Knative Authors
```

mattmoor replied:
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2020 The Knative Authors
```

mattmoor replied:
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2020 The Knative Authors
```

mattmoor commented on pkg/kn/commands/subscription/describe_test.go:
```diff
@@ -0,0 +1,53 @@
+// Copyright © 2020 The Knative Authors
```
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2020 The Knative Authors
```

mattmoor replied:
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2020 The Knative Authors
```

mattmoor commented on pkg/kn/commands/subscription/flags.go:
```diff
@@ -0,0 +1,137 @@
+// Copyright © 2020 The Knative Authors
```
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2020 The Knative Authors
```

mattmoor replied:
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2020 The Knative Authors
```

mattmoor replied:
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2020 The Knative Authors
```

mattmoor commented on pkg/kn/commands/subscription/list.go:
```diff
@@ -0,0 +1,73 @@
+// Copyright © 2020 The Knative Authors
```
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2020 The Knative Authors
```

mattmoor replied:
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2020 The Knative Authors
```

mattmoor replied:
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2020 The Knative Authors
```

mattmoor commented on pkg/kn/commands/subscription/list_test.go:
```diff
@@ -0,0 +1,52 @@
+// Copyright © 2020 The Knative Authors
```
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2020 The Knative Authors
```

mattmoor replied:
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2020 The Knative Authors
```

mattmoor commented on pkg/kn/commands/subscription/subscription.go:
```diff
@@ -0,0 +1,68 @@
+// Copyright © 2020 The Knative Authors
```
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2020 The Knative Authors
```

mattmoor replied:
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2020 The Knative Authors
```

mattmoor replied:
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2020 The Knative Authors
```

mattmoor commented on pkg/kn/commands/subscription/subscription_test.go:
```diff
@@ -0,0 +1,81 @@
+// Copyright © 2019 The Knative Authors
```
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2019 The Knative Authors
```

mattmoor replied:
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2019 The Knative Authors
```

mattmoor commented on pkg/kn/commands/subscription/update.go:
```diff
@@ -0,0 +1,102 @@
+// Copyright © 2020 The Knative Authors
```
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2020 The Knative Authors
```

mattmoor replied:
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2020 The Knative Authors
```

mattmoor replied:
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2020 The Knative Authors
```

mattmoor commented on pkg/messaging/v1beta1/subscriptions_client.go:
```diff
@@ -0,0 +1,189 @@
+// Copyright © 2020 The Knative Authors
```
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2020 The Knative Authors
```

mattmoor replied:
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2020 The Knative Authors
```

mattmoor replied:
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2020 The Knative Authors
```

mattmoor commented on pkg/messaging/v1beta1/subscriptions_client_mock.go:
```diff
@@ -0,0 +1,119 @@
+// Copyright © 2020 The Knative Authors
```
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2020 The Knative Authors
```

mattmoor replied:
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2020 The Knative Authors
```

mattmoor replied:
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2020 The Knative Authors
```

mattmoor reviewed (commented):
<!--copyright-bot-->

mattmoor reviewed (commented):
<!--copyright-bot-->

maximilien reviewed (approved):
/retest

maximilien commented on pkg/kn/commands/subscription/flags.go:
```diff
@@ -0,0 +1,137 @@
+		{Name: "Namespace", Type: "string", Description: "Namespace of the Subscription", Priority: 0},
+		{Name: "Name", Type: "string", Description: "Name of the Subscription", Priority: 1},
+		{Name: "Channel", Type: "string", Description: "Channel of the Subcription", Priority: 1},
+		{Name: "Subscriber", Type: "string", Description: "Subscriber of the Subscription", Priority: 1},
```
Subscriber or subscription is a mouthful :) But not sure there is better

navidshaikh replied:
thats for the list column, however while we are on it, I wanted to add an alias `subs` for `subscription` command group. I think we also had a discussion earlier to create a spreadsheet or similar document for the alias proposal? I think we should do that for the coming release ?

maximilien reviewed (changes requested):
Any chance for an e2e test for this?

mattmoor commented on pkg/kn/commands/subscription/create_test.go:
```diff
@@ -0,0 +1,82 @@
+/*
```
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

/*
```

navidshaikh replied:
@mattmoor : followed the license suggestion and added to create_test.go in same format, it sill complains though?

mattmoor replied:
I believe (without debugging) that it's complaining because the header comment doesn't match the `boilerplate.go.txt` file (likely somewhere under `./hack/`).  There are a few subtle differences between what you added and what it's suggesting, which may be triggering it.

navidshaikh replied:
yes, the one in PR carries this additional character in license `©`
```
-Copyright © 2020 The Knative Authors
+Copyright 2020 The Knative Authors
```

mattmoor commented on pkg/kn/commands/subscription/delete_test.go:
```diff
@@ -0,0 +1,54 @@
+/*
```
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

/*
```

mattmoor commented on pkg/kn/commands/subscription/describe_test.go:
```diff
@@ -0,0 +1,59 @@
+/*
```
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

/*
```

mattmoor commented on pkg/kn/commands/subscription/list_test.go:
```diff
@@ -0,0 +1,71 @@
+/*
```
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

/*
```

mattmoor commented on pkg/kn/commands/subscription/subscription_test.go:
```diff
@@ -0,0 +1,145 @@
+// Copyright © 2019 The Knative Authors
```
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Copyright © 2019 The Knative Authors
```

mattmoor commented on pkg/kn/commands/subscription/update_test.go:
```diff
@@ -0,0 +1,74 @@
+/*
```
<!--copyright-bot-->
Missing copyright header:
```suggestion
/*
Copyright 2020 The Knative Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

/*
```

mattmoor reviewed (commented):
<!--copyright-bot-->

navidshaikh commented on docs/cmd/kn_subscription.md:
```diff
@@ -0,0 +1,35 @@
+
+* [kn](kn.md)	 - kn manages Knative Serving and Eventing resources
+* [kn subscription create](kn_subscription_create.md)	 - Create an event subscription
+* [kn subscription delete](kn_subscription_delete.md)	 - Delete a subscription
```
changed to use `a subscription` with all CRUD commands

dsimansk commented on pkg/messaging/v1beta1/subscriptions_client.go:
```diff
@@ -0,0 +1,189 @@
+}
+
+// Build returns the Subscription object from the builder
+func (c *SubscriptionBuilder) Build() *v1beta1.Subscription {
+	return c.subscription
```
Just a cosmetic one.
```suggestion
func (s *SubscriptionBuilder) Build() *v1beta1.Subscription {
	return s.subscription
```

navidshaikh commented:
/retest


     ERROR: timeout waiting for pods to come up
     eventing-controller-56cdd9bf45-cmttb    1/1   Running            0     5m42s
     eventing-webhook-8665cc79c6-sfr9d       0/1   CrashLoopBackOff   6     5m42s

rhuss reviewed (commented):
Looks good to me and is a straight forward implementation for subscriber CRUD handling.

There are two things I would still like to discuss though:

* I know that the options UI mimic 1:1 the fields in the resources, but I wonder whether we should do this here as well. I.e. the target for where events go a called a _sink_ for sources and triggers, but _subscriber_ for subscriptions although they are all of the same types and share the same mappings, that we call "sink mappings". I tend to name the 'subscriber' here also --sink or at least make that option an alias. tbh I have no idea why this difference, but from a UI POV this is more than unfortunate. The same, but to a minor degree, for the difference between a `--dead-letter-sink` (with sink suffix) and `--reply` (without sink suffix, but still a sink). Ideally either both or non would have a sink prefix.

* Should we settle on the `v1` API version of subscription? I would say yes, as I think that for initial support we should use the highest available API version.

rhuss commented on docs/cmd/kn_subscription_create.md:
```diff
@@ -0,0 +1,46 @@
+      --dead-letter-sink string   Addressable sink for events. You can specify a broker, Knative service or URI. Examples: '--dead-letter-sink broker:nest' for a broker 'nest', '--dead-letter-sink https://event.receiver.uri' for an URI with an 'http://' or 'https://' schema, '--dead-letter-sink ksvc:receiver' or simply '--dead-letter-sink receiver' for a Knative service 'receiver'. If prefix is n ...
+  -h, --help                      help for create
+  -n, --namespace string          Specify the namespace to operate in.
+      --reply string              Addressable sink for events. You can specify a broker, Knative service or URI. Examples: '--reply broker:nest' for a broker 'nest', '--reply https://event.receiver.uri' for an URI with an 'http://' or 'https://' schema, '--reply ksvc:receiver' or simply '--reply receiver' for a Knative service 'receiver'. If prefix is not provided, it is considered as a Knative s ...
```
As @grantr pointed out, the original difference was that a _sink_ don't return a CE but a _subscriber_ does. According to https://github.com/knative/eventing/blob/master/docs/spec/data-plane.md , this definition does not really shine through (i.e. a sink _can_ also return a value). Also `kn trigger` uses `--sink` but fills in a `subscriber` field internally.

So I tend to the following solution:

* Use `--sink` also for subscriptions
* Use `--reply-sink` instead of only `--reply` to make is consistent with `--dead-letter-sink`

rhuss replied:
After some discussion in https://knative.slack.com/archives/C9JP909F0/p1601309985065600 I like that suggestion best:

* Use `--sink` also for subscriptions
* Use  `--sink-reply` and `--sink-dead-letter` to group all sink typed options together

rhuss replied:
I put the discussion about the proper option names to today's wg call.

navidshaikh replied:
I am fine changing the flag names to better align with other command's flags, should this also go to eventing to change these  fields name in the SPEC to better align ? 

> Use --sink-reply and --sink-dead-letter to group all sink typed options together

I wonder if `sink` should be a suffix than prefix here, this way we dont deviate from the fields in spec (only 5 flags here, so I dont think grouping would be useful enough).

rhuss replied:
well, for `Reply` we would deviate anyway.  So having those together in the list might make sense ?

navidshaikh commented on docs/cmd/kn_subscription_create.md:
````diff
@@ -0,0 +1,46 @@
+### Options
+
+```
+      --channel string            Specify the channel to subscribe to, in the format '--channel Group:Version:Kind:Name'. You can use channel type aliases from kn config with this flag. You can also refer inbuilt channel type InMemoryChannel using alias 'imc' like '--type imc:CHANNEL_NAME'. Examples: '--channel messaging.knative.dev:v1alpha1:KafkaChannel:k1' for specifying explicit Group:Version: ...
````
> Is there are a way to reference a channel only by its name ? I.e. is there kind of a "default" channel type that would be picked when no GVK coordinates are specified.

No. We expect to have an explicit channel type mentioned.

(for channel CRUD, we defer the type of channel when creating to messaging layer default, and require only name of channel to create)

navidshaikh replied:
> However I wonder whether we can already come up with a default set of mappings that are already covered by eventing, like the kafka broker that you mention here.

sure, this came up during the channel CRUD implementation as well, but since kafka source isnt a native install of eventing, we didnt assume the GVK of same to map to a default alias.

rhuss replied:
yeah, I guess also that we then again have issues with the proper version. E.g. I would reserve `kafka:` for `v1` (or the "latest") available version (so that the standard prefix can grow with the version)

navidshaikh commented on pkg/kn/commands/flags/sink.go:
```diff
@@ -33,16 +33,30 @@ type SinkFlags struct {
 
-func (i *SinkFlags) Add(cmd *cobra.Command) {
+func (i *SinkFlags) Add(cmd *cobra.Command, fname ...string) {
+	// FIXME: Quick hack to avoid changing the args at all places,
```
:+1: I will go with second suggestion.

navidshaikh commented on pkg/messaging/v1beta1/subscriptions_client.go:
```diff
@@ -0,0 +1,191 @@
+type KnSubscriptionsClient interface {
+
+	// GetSubscription returns a Subscription by its name
+	GetSubscription(name string) (*v1beta1.Subscription, error)
```
this should help `kn` to work with earlier eventing installs ?
( IIRC, we shifted to serving v1 when the stored version changed ?)

navidshaikh commented:
This is ready for another round of review, fixed the review comments and here are some notable changes:

- Flag names revisited:
  -   -s, --sink string
  -  --sink-dead-letter string
  -  --sink-reply string

- Subscription list column name revisited

- Aliases:
  - subscriptions
  - sub

- Added shared lib:
  knative.dev/client/lib/printing

rhuss reviewed (commented):
cool, thanks ! 

/lgtm
/approve

navidshaikh commented:
/retest

     ERROR: timeout waiting for pods to come up

navidshaikh commented:
/retest

rhuss commented:
/retest

navidshaikh commented:
/retest


looks like https://github.com/knative/eventing/issues/4165 is flaking the CI job

navidshaikh commented:
/hold

adding IMC as default channel if no prefix is given 
see https://knative.slack.com/archives/CE4MVFVAQ/p1601473090002600

navidshaikh commented on CHANGELOG.adoc:
```diff
@@ -18,12 +18,16 @@
 
 | 🎁
-| Add support for service initialScale via `--scale-init` flag
+| Add support for service initialscale via `--scale-init` flag
```
```suggestion
| Add support for service initialScale via `--scale-init` flag
```

navidshaikh commented on pkg/kn/flags/channel_types.go:
```diff
@@ -79,3 +86,35 @@ func (i *ChannelTypeFlags) Parse() (*schema.GroupVersionKind, error) {
+func (i *ChannelRef) Parse() (*corev1.ObjectReference, error) {
+	parts := strings.Split(i.Cref, ":")
+	switch len(parts) {
+	// TODO: Check if we want an implicit assumed channel type to be of IMC
+	// case 1:
```
@rhuss : see ;-)

rhuss commented on pkg/kn/flags/channel_types.go:
```diff
@@ -93,21 +93,24 @@ func (i *ChannelRef) Add(f *pflag.FlagSet) {
+			"You can refer channel type aliases from kn config with this flag. "+
+			"You can also refer inbuilt channel type aliases 'imcv1beta1' or 'imc'. "+
+			"Examples: '--channel messaging.knative.dev:v1alpha1:KafkaChannel:k1' for specifying explicit Group:Version:Kind:Name. "+
+			"If a prefix is not provided, it is considered as Channel(messaging.knative.dev/v1beta1).")
```
```suggestion
			"If no prefix is provided, the default channel is used.")
```

rhuss replied:
Or maybe even better rephrase to mention the default channel first, like in:

```
Specify the channel to subscribe to. For the default channel, just use the name (e.g. "mychannel"). A mapped channel type like "imc" can be used as a prefix (e.g. "imc:mychannel"). Finally you can specify the full coordinates to the referenced channel with Group:Version:Kind:Name (e.g. "messaging.knative.dev:v1alpha1:KafkaChannel:mychannel")
```

navidshaikh replied:
thanks! I've used the second suggestion.

rhuss reviewed (commented):
Looks good, some suggestion for a more condensed wording in the help message. I would not mention any concrete build int mappings here but defer that more or less to the channel documentation,

rhuss commented:
/retest

navidshaikh commented:
/retest

       result_collector.go:75: ERROR: Error: ReconcileIngressFailed: Ingress reconciliation failed

rhuss commented:
/lgtm

navidshaikh commented:
/unhold

rhuss commented:
/retest
